In [27]:
import numpy as np
from statsmodels.stats.contingency_tables import mcnemar

# y = true labels; pa = Linear-SVC predictions; pb = ensemble predictions
# (pooled out-of-fold, same documents in the same order)
ccat_50_dfse = '''23 41 38 34 39 10  1 30 31 11 28 48 24 41 11  1 41 39 45 20 31 29 17 28
 16 13 25 39  6 15 22 48  5  5 19  4 48 19 27 32 11 36  1 15  2 28 23 27
  4 43 35  7 48 48 29 44 39 47 36 22 45 26 20 12 28 26 38 12 31 18 32 47
 24  1  4 14 32 38 13 49 22 10 42 30 34 14 40 28 44 47 23  7 47 39 47 31
 29  6 11 15 26 35  2 19 21  7 14 14 15 16  5 44  9 23  4 14  1 37 31 32
 24 43  4 12 11 18  0 37 30  9 44 17 37  4 16 21 48 26 46 48 35 45 46 43
 36 37 11 35 47 44 42 20  7  7 36  1 16 37 28  5 22 17 35 29 22 19 10 29
 45  0 12 49 40 12 44 19 39 33 26 16 42 31 46 31 46 46 33  8 32 42 47 39
 36 44 24  6 23 49 10 11 25  4  6 33  0 29 40 12 31  5  2 32  8 22 12 28
 17 28 36 19 18 10 33 47  4 37 20 43 29 43 37 25 25 39 37 44 13  6 26 42
 10 21 48  0 30  3  6 27 39 42 43  1 44  4 47 12 44  3  9  6  0 13 49 47
 40 37 28  1 38 19 40 38 38 22 11 41 48 17 18 17 26 24 21  4 40 39 19  9
 40 10 47  4 36 35 49 48  5 18 33 11 29 23 26 39 19 26 17 10  6 13  6 35
 46  6  9 12  0  6 28 11 32  3 21 23 47 34 35 30 10 22 23 10 37 18 37 47
 11 14 20 30 46  1 14 33 21 43 41 38 19 11 41 11 17 27 22  6 23 31 36 15
  1 17 28 30  0  3  6 16 30 19 43 11 47 33  7  4 25 37 10  6 41  3 39 33
 27 49 38 41 31 10 31 10  6 25 37 28  0 31 31 30  8 41 14 27  7 31 15 48
 27 14  2 35  3  3  1 34 24 39 11 12  8 28 13 35 37 44 35 33 21 16 25 15
 28 31 47  9 44  0  7 30 17 38 32 29 19 34 30  0 30  9 16 26 42 21 27 20
 32 21 38 40 39 29 14 22 28 32 25 29 30 15 11 16  7 28 18 47 43  6 35 43
 19  0 15 22 33 24 37 42 42 29 27 24  6 25 13 22 49 43 18 15 46  2 38  0
  5 30 20 34 27  7 10  6 47 20  0  4  1 43 36 24 44 21 34 42 45  0 35 24
 33 46  2  7 34 29 20  8 36 16  1 18  7 27  1 47 13 47 43 13 36  1 38 38
 20 11 37 36 46  9  9 11 49 34 48 23  3 11 17 40 35 21 17  4 46 35 42 16
  5  1 44  6 11 34  8 48 17 41 20 35 45  0 16 11 17 12  5 38 15 13 14  3
 26 25 30 10 42 49 15 39 21 44 24 16 32 37 49 40 46  9  5 34 38 27 11 47
 43  3 49 32 14  4 40 35 49 12  4 44 37  9 25 12 49 43 42 48  7  9 26 46
 16 24  7 25  5  5 13 20 49 29 16 12 25 21 49 26 13 43 45 36 22 25 27  9
 31  2 20 12 39  9 37 22 12 38 30 12 21  3 41 40 17  9  7 42 25 31 39 17
 22 26 32 22 13  3 32  0  3 46 35 42 34 36 19 24 13 16 29 18 37  7 13  0
 27 37 44 18  1 31 19  9  5 21  6 46 47  4  7 15 44  1  1  8 40 29 42  1
  7 24 46 31 20  3 26  4 32 43  4 35 46 23 44  4 28  5 17 48 44  2 37 22
 45 24 24 40 18 46 34 20 12 15 35  7 14  0 23 28 34 47 41 17  2 36  8 49
 37 25 28 21 42 16  7 49 30 10  3  8 39 29 36 10 11 11  3  0 48 49 22 17
 47 31 12 31 49 41 49 27 17 12 41 14 29 21 33 21  9 16 31 19 27 37 28 23
 13 44 16 20 35 26 42 12 44 25 42 17 17  9 29 37 39 37 15 32 40 14 18 16
  4 30 27 49 35 40 16 18  5  5  2 10 29 37  4 16 47  7  2 11 18 47 27  1
 33 42 13 11 42 18  9 17 39 33 15 36 46 15 47 38 14  7 31  5  4 33 17 35
 15 11  9  2 38 49 37 29 20 32 24 36 34 46  6 32 10  1  8  8 16  9 25 43
 37 40 24 46 19 39  8 38 30 32 19 49 27 20 11 31 43 26 38  2 27 49 13 39
 23 30  7 46 21 15 29 33 40 49 28 25 30  2 36 20  3 33 48 46 10 42 14 13
 21 24 23 46 18 23 16 18 40 30 45 40 26 22 33 41'''.replace('\n','')
ccat_50_dfse = np.array([int(i) for i in ccat_50_dfse.split()])
#a_ok, b_ok = (pa == y), (pb == y)
#table = [[np.sum(a_ok & b_ok),  np.sum(a_ok & ~b_ok)],
#         [np.sum(~a_ok & b_ok), np.sum(~a_ok & ~b_ok)]]
#print(mcnemar(table, exact=True))   # two-sided p; halve it if pb gains over pa

In [28]:
ccat_50_dfse_true = '''23 41 38 34 39 10  1 30 31 11 28 48 24 41 11  1 46 39 45 20 31 29 17 28
 16 18 25 39  6 15 22  7  5  5 19  4 48 19 27 32 11 36  1 15  2 28 23 27
  4 43 35  7 48  7 29 44 39 47 36 22 45 26 20 12 28 26 38  6 31 18 32  8
 24  1 23 14 32 38 13 34  5 10 23 30 34 14 40 28 44 24 23 48 47 39 47 31
 29  6 11 15 26  0 17  8 21 48  3  3 15 16  5 44  0 23  4  3  1 45 31 32
 24 43 23 12 11 18  0 37 33  9 44 44 37  4 24 21 48 26 25 48 35 42 46 14
 36 37 14  9  1 44 42 20  7 48 36  1 16 45 28  5 22 21 35 29 22 19 10 29
 37  0 12 34 40 12 44 19  5 33 26 16 42 31 31 31 46 25 33  8 32 42  2 39
 36 44 24 26 23 43 10 11 25  4  6 33  0 29 40 12  5  5  2 32 38 22 26 28
  2 28 22 19 18 10 33 47  4 37 20 43 29 32 37 25 25 39 45 44 13  6 26 42
 10 21 48  0 30 34  6 27 39 45 43 18 44  4 21 12 44 43  9  6  0 13 49 19
 40 45 28  1 38 19 40 38 38 22 11 41 48 17 13 17 26 24 21  4 40 39  8  9
 40 10 47  4 36 35 49  5 22 18 33 11 29 23 26 39 19 26 17 10  6 13  6 35
 25  6  9 12  0  6 28 11 32 49 21 23 19 34 21 23 10 36 23 10 37 18 45 21
 11 34 20 30  9  1 43 33 21 20 41 38 19 11 41 11 17 34 22  6 23 25 36 15
  1 17 28 30  0  3 12 19 30 19 14 11  8 33 48 30 46 49 10  6 25  3 39 48
 27 49 36 41 31 10 31 10  6 25 37 28 24 39 31 30  8 41 14 27  7 31 15 48
 27 14 17 35 34  3  1 34 16 39 42 12  8 28 18 35 37 44 35 33  2 16 41 15
 28 31 47  9 44 14  7 30 17 38 32 29  8 11 34  0 30  9 16 26 42 21 27 20
 32  8 38 40 41 29 14 22 28 32 25 29 30 15 49 47  7 28 18 47 37  6 35 45
 19  0 15 22 33 24 37 42 42 29 27 19  6 41 18 22 34 43 18 15 46  2 38  0
  5 30 20 43 27  7 10  6 47 20  0  4 13 49 36 24  2 21 49 42 45  0 35 24
 33 46  2 48 43 38 20  8 36 16  1 18  7 27  1 47 13 47  3 13 36 18 38 38
 20 27 45 36 46  0 46 11 34  3 48 23  3 14 17 40 35 21 24 23  9 35 42 16
  5  1 44  6 49 34  8 48 17 41 20 35 43  0 16 42 17 12  5 38 15 13 14 14
 26 25 30 10 37 49 15  9 21  2 24 16 32 34 49 40 46 41  5 43 38 27 14 47
 14  3 43 32 14  4 40 35 49 12  4 44 42  9 25 12 43  3 42 48  7  9 26 46
 16 24  7 25  5  5 13 20 49 29 16 12 25 33 49 26 13 43 45 22 22 25 27 46
  9  2 20 12 39  9 37 36 12 38 30 12 21  3 41 40  2  9  7 42 25 31 39  2
 22 26 32 22 13 14 32  0  3 41 35 42 34 36 19 24 13 16 29 18 37  7 13  0
 27 34 47 18  1 31 19 46  5 15  6 46  8  4  7 15 44  1  1  8 40 29 45  1
 48 24 41 31 20  3 26  4 32 43  4 35 41 30 44  4 28  5 17  7 44  2 45 22
 45 24 24 40 13 46 43 20 12 15 35  7  3  0 23  8 43 47 41 17  2 36  8 34
 45  9 28 21 45 16  7 49 30 10  3  8 39 47 36 10 49 11  3  0 48 49 22 47
 47 31 12 31 37 41 49 27  2 12 41 14 29 21 33 21  9 19 33 19 27 37 28 23
 18 44 16 20 35 26 45 12 44 25 45 17 17  9 29 37 39 37 15 32 40 43 18 16
  4 30 27  3 35 40 16 18 36  5  2 10 29 37  4 16 47  7  2 27 18 47 49  1
 33 42 13 11 42 18  9 17 39 33 15 36 46 15 47 17 14  7 31  5  4 33 17 35
 15 11 46  2 38 11 37 29 20 32 21  5 34 46  6 32 10  1  8  8  8  9 25 43
 42 40 24 46 19 39  8 38 30 32 19 34 27 20 11 22  3 26 38 17 27 14 13 39
 23 30  7 46 21 15 29 33 40 49 28 25 30  2 36 20  3 33 48 46 10 42 14 13
  2 24 23 31 13  4 16 13 40 23 45 40 26 22 33 41'''.replace("\n","")
ccat_50_dfse_true = np.array([int(i) for i in ccat_50_dfse_true.split()])

In [29]:
ccat_50_dfse_linear_svc = '''23 41 38 34 39 10  1 30 31 11 28 48 24 41 11  1 41 39 37 20 31 29 17 28
 16 13 25 39  6 15 22 48  5  5 19  4 48 19 27 32 11 36  1 15  2 28 23 27
 30 43 35  7 48  7 19 44 39 47 36 22 42 26 20 12 28 26 38 12 31 18 32 21
 24  1  4 14 32 38 13 49  5 10 42 30 34 14 40 28 44 47 23  7 47 39 47 31
 29  6 11 15 26 35  2 21 21  7 14 14 15 16  5 44  9 23  4  3  1 37 31 32
 24 43  4 12 11 18  0 37 30  9 44 17 37  4 16 21 48 26 41 48 35 45 46 43
 36 37 11 35  2 44 42 20  7  7 36  1 31 37 28  5 22 24 35 29 22 19 10 29
 45  0 12 49 40 12 44 19  5 33 26 16 42 31 41 31 46 46 33  2 32 42 21 39
 36 44 24  6 23 49 10 11 25  4  6 33  0 29 40 12 41  5  2 32  8 22 12 28
 17 28 22 19 18 10 33 47  2 45 20 49 29 32 45 25 25 39 37 44 13  6 26 42
 10 21 48  0 30  3  6 27 39 42 43  1 44 30 47 12 44  3  9  6 40 13 34 31
 40 37 28  1 38 19 40 38 31 22 11 41 48 44 18 17 26 24 21  4 40 39 10  9
 40 10 47  4 36 35 49  7  5 18 33 11 29 23 26 39 19 26 17 10  6 13 12 35
 46  6  9 12  0  6 28 11 32  3 21 23 47 34 35 26 10  5 23  0 37 18 37 47
 11 14 20 30 46  1 49 33 21 43 41 38 19 11 41 11 17 34 22  6 23 31 36 15
  1 17 28 30  0  3 12 16 30 19 43 11 21 33 48 23 25 45 10  6 41  3 39 36
 27 49 38 41 31 10 31 10  6 25 37 28  0 31 31 30 21 41 14 27  5 31 15 48
 27 49  2 35  3  3  1 34 24 39 42 12  8 28 13 35 37 44 35 33 21 16 25 15
 28 31 47  9 44  0  7 30 17 38 32 38  8 34 30  0 30  9 16 26 42 21 27 20
 32 32 29 40  5 29 14 22 28 32 25 29 30 15 11 16  7 28 18 47 43 12 35 43
 19  0 15 22 33 24 37 42 42 29 27 24  6 25 13 22 49 43 18 15 46  2 38  0
  5 30 20 34 27  7 10  6 47 20  0  4  1 49 36 24 44 21 43 42 45  0 35 24
 33 46  2  7 34 29 20  8 36 16  1 18  7 27  1 47  1 47 49 13 36  1 38 38
 20 11 37 36 46  9  9 11 49  3 48 23  3 11 17 40 35 21 17 30 41 35 42 16
  5  1 44  6 11 34  8 48 17 41 20 35 45  0 16 11 17 12  5 38 15 13 14  3
 26 25 30 10 42 34 15 39 21 31 24 16 32 45 14 40 46 31  5 34 38 27 11 47
 43  3 14 32 14  4 40 35 14 12  4 44 45  9 25 12 49  3 42 48  7  9 26 46
 16 24  7 25  5  5 13 20 49 29 16 12 25 21 49 26 13 43 45  5 22 25 27  9
 33  2 20 12 39  9 37 22 12 38 30 12 21 49 41 40 17  9  7 42 25 31 39 17
 22 26 32 22 13  3 32  0  3 41 35 42 34 36 19 24  1 16 29  1 37  7 13  0
 27 45 44 18  1 40 19  9  5 15 12 41 47  4  7 15 44  1  1  8 40 29 42 18
  7 24 25 31 20  3 26  4 32 49  4 35 41 23 44  4 28  5 17 48 44  2 37 22
 45 24 24 40 18 46 34 20 12 15 35  7 14  0 23 28 34 21 41 17  2 36  8 49
 37 25 28 21 45 16  7 49 30 10  3 21 39 38 36 10 11 11  3  0 48 49 22 17
 47 31 12 31 49 41 49 27 17 12 41 14 29 21 33 21  9 16 31 19 27 37 28 23
 13 44 16 20 35 26 42 12 44 25 42 17 17  9 29 37 39 37 15 32 40 14 18 16
  4 30 27 49 35 40 16 18  5  5  2 10 29 37  4 16 16  7  2 27 18 47 34  1
 33 42 13 11 42 18  9 17 39 33 15 36 46 15 47  8 14  7 31  5  4 33 17 35
 15 11 41  2 38 49 37 29 20 32 21 36 34 46 12 32 10  1  8  8 21  9 25 43
 37 40 24 46 19 39  8 38 30 32 19 11 27 20 11 31 43 26 38  2 27 49 13 39
 23 30  7 46 21 15 29 33 40 11 28 25 30 17 36 20  3 33 48 41 10 11 14  1
  2 24 23 41 18 23 16 18 40 30 37 40 26  5 33 41'''.replace("\n","")
ccat_50_dfse_linear_svc = np.array([int(i) for i in ccat_50_dfse_linear_svc.split()])

In [32]:
ccat_50_dfse_linear_svc_true = '''23 41 38 34 39 10  1 30 31 11 28 48 24 41 11  1 46 39 45 20 31 29 17 28
 16 18 25 39  6 15 22  7  5  5 19  4 48 19 27 32 11 36  1 15  2 28 23 27
  4 43 35  7 48  7 29 44 39 47 36 22 45 26 20 12 28 26 38  6 31 18 32  8
 24  1 23 14 32 38 13 34  5 10 23 30 34 14 40 28 44 24 23 48 47 39 47 31
 29  6 11 15 26  0 17  8 21 48  3  3 15 16  5 44  0 23  4  3  1 45 31 32
 24 43 23 12 11 18  0 37 33  9 44 44 37  4 24 21 48 26 25 48 35 42 46 14
 36 37 14  9  1 44 42 20  7 48 36  1 16 45 28  5 22 21 35 29 22 19 10 29
 37  0 12 34 40 12 44 19  5 33 26 16 42 31 31 31 46 25 33  8 32 42  2 39
 36 44 24 26 23 43 10 11 25  4  6 33  0 29 40 12  5  5  2 32 38 22 26 28
  2 28 22 19 18 10 33 47  4 37 20 43 29 32 37 25 25 39 45 44 13  6 26 42
 10 21 48  0 30 34  6 27 39 45 43 18 44  4 21 12 44 43  9  6  0 13 49 19
 40 45 28  1 38 19 40 38 38 22 11 41 48 17 13 17 26 24 21  4 40 39  8  9
 40 10 47  4 36 35 49  5 22 18 33 11 29 23 26 39 19 26 17 10  6 13  6 35
 25  6  9 12  0  6 28 11 32 49 21 23 19 34 21 23 10 36 23 10 37 18 45 21
 11 34 20 30  9  1 43 33 21 20 41 38 19 11 41 11 17 34 22  6 23 25 36 15
  1 17 28 30  0  3 12 19 30 19 14 11  8 33 48 30 46 49 10  6 25  3 39 48
 27 49 36 41 31 10 31 10  6 25 37 28 24 39 31 30  8 41 14 27  7 31 15 48
 27 14 17 35 34  3  1 34 16 39 42 12  8 28 18 35 37 44 35 33  2 16 41 15
 28 31 47  9 44 14  7 30 17 38 32 29  8 11 34  0 30  9 16 26 42 21 27 20
 32  8 38 40 41 29 14 22 28 32 25 29 30 15 49 47  7 28 18 47 37  6 35 45
 19  0 15 22 33 24 37 42 42 29 27 19  6 41 18 22 34 43 18 15 46  2 38  0
  5 30 20 43 27  7 10  6 47 20  0  4 13 49 36 24  2 21 49 42 45  0 35 24
 33 46  2 48 43 38 20  8 36 16  1 18  7 27  1 47 13 47  3 13 36 18 38 38
 20 27 45 36 46  0 46 11 34  3 48 23  3 14 17 40 35 21 24 23  9 35 42 16
  5  1 44  6 49 34  8 48 17 41 20 35 43  0 16 42 17 12  5 38 15 13 14 14
 26 25 30 10 37 49 15  9 21  2 24 16 32 34 49 40 46 41  5 43 38 27 14 47
 14  3 43 32 14  4 40 35 49 12  4 44 42  9 25 12 43  3 42 48  7  9 26 46
 16 24  7 25  5  5 13 20 49 29 16 12 25 33 49 26 13 43 45 22 22 25 27 46
  9  2 20 12 39  9 37 36 12 38 30 12 21  3 41 40  2  9  7 42 25 31 39  2
 22 26 32 22 13 14 32  0  3 41 35 42 34 36 19 24 13 16 29 18 37  7 13  0
 27 34 47 18  1 31 19 46  5 15  6 46  8  4  7 15 44  1  1  8 40 29 45  1
 48 24 41 31 20  3 26  4 32 43  4 35 41 30 44  4 28  5 17  7 44  2 45 22
 45 24 24 40 13 46 43 20 12 15 35  7  3  0 23  8 43 47 41 17  2 36  8 34
 45  9 28 21 45 16  7 49 30 10  3  8 39 47 36 10 49 11  3  0 48 49 22 47
 47 31 12 31 37 41 49 27  2 12 41 14 29 21 33 21  9 19 33 19 27 37 28 23
 18 44 16 20 35 26 45 12 44 25 45 17 17  9 29 37 39 37 15 32 40 43 18 16
  4 30 27  3 35 40 16 18 36  5  2 10 29 37  4 16 47  7  2 27 18 47 49  1
 33 42 13 11 42 18  9 17 39 33 15 36 46 15 47 17 14  7 31  5  4 33 17 35
 15 11 46  2 38 11 37 29 20 32 21  5 34 46  6 32 10  1  8  8  8  9 25 43
 42 40 24 46 19 39  8 38 30 32 19 34 27 20 11 22  3 26 38 17 27 14 13 39
 23 30  7 46 21 15 29 33 40 49 28 25 30  2 36 20  3 33 48 46 10 42 14 13
  2 24 23 31 13  4 16 13 40 23 45 40 26 22 33 41'''.replace("\n","")
ccat_50_dfse_linear_svc_true = np.array([int(i) for i in ccat_50_dfse_linear_svc_true.split()])

In [33]:
a_ok, b_ok = (ccat_50_dfse == ccat_50_dfse_true), (ccat_50_dfse_linear_svc == ccat_50_dfse_linear_svc_true)
table = [[np.sum(a_ok & b_ok),  np.sum(a_ok & ~b_ok)],
         [np.sum(~a_ok & b_ok), np.sum(~a_ok & ~b_ok)]]
print(mcnemar(table, exact=True))   # two-sided p; halve it if pb gains over pa

pvalue      0.0030653176696087397
statistic   21.0


In [34]:
print("both right:", np.sum(a_ok & b_ok))
print("only pa right:", np.sum(a_ok & ~b_ok))   # should be 46 if pa = ensemble
print("only pb right:", np.sum(~a_ok & b_ok))   # should be 21 if pb = Linear SVC
print("acc pa:", a_ok.mean(), "acc pb:", b_ok.mean())

both right: 747
only pa right: 46
only pb right: 21
acc pa: 0.793 acc pb: 0.768


On a 1,000-document CCAT_50 test fold, the ensemble and Linear SVC agreed on 93% of documents. The ensemble corrected 46 of the baseline's 232 errors while introducing 21 new ones (exact McNemar p = 0.003; accuracy gain 2.5 points, 95% CI roughly 0.9 to 4.1).

What it tells you
The models mostly agree. They give the same verdict on 933 of 1,000 documents (747 both right, 186 both wrong), so only 67 documents (6.7%) decide the comparison.
The ensemble's net effect is positive. It fixed 46 of Linear SVC's 232 errors (about 20%) and introduced 21 new ones, a net gain of 25 documents, or 2.5 points. That matches the 2.5-point gap in your Table 3 (79.04 vs 76.54), so this fold is representative of the means.
The gain is statistically reliable on this fold. The exact two-sided McNemar p is 0.0031. The 95% confidence interval for the accuracy gain is about 2.5 ± 1.6 points, so roughly 0.9 to 4.1 points. The interval excludes zero, but the gain could be as small as one point.
The ceiling is shared. 186 documents (18.6%) are misclassified by both models. The ensemble does nothing about them, because the two use the same 448-dimensional DFSE features. That limits what any kernel reweighting can achieve and fits your earlier observation that the gain is modest.

In [36]:
ccat_10_dfse = '''0, 4, 3, 8, 4, 9, 1, 6, 8, 4, 0, 8, 6, 5, 9, 8, 4, 6, 4, 5, 2, 6, 0, 5, 7, 2, 9, 7, 5, 9, 9, 5, 0, 4, 9, 5, 0, 3, 4, 1, 2, 4, 6, 5, 4, 7, 9, 0, 0, 9, 8, 1, 1, 9, 7, 8, 0, 9, 3, 4, 1, 5, 5, 9, 8, 6, 8, 5, 8, 7, 3, 5, 8, 4, 6, 1, 0, 0, 9, 1, 4, 8, 3, 0, 5, 4, 0, 4, 9, 4, 7, 4, 5, 9, 6, 2, 5, 6, 3, 0, 6, 2, 8, 1, 2, 9, 7, 7, 3, 0, 1, 1, 3, 2, 4, 8, 6, 6, 9, 5, 9, 0, 0, 7, 1, 4, 6, 2, 7, 2, 4, 1, 4, 1, 9, 1, 9, 2, 1, 8, 3, 6, 9, 0, 2, 8, 7, 2, 2, 0, 8, 7, 5, 7, 9, 8, 1, 7, 9, 5, 6, 6, 9, 7, 1, 7, 9, 8, 1, 8, 6, 5, 2, 6, 6, 5, 4, 1, 0, 8, 3, 4, 7, 4, 0, 7, 7, 5, 9, 2, 2, 0, 0, 9, 2, 1, 7, 4, 6, 3'''.replace('\n','')
ccat_10_dfse = np.array([int(i) for i in ccat_10_dfse.split(',')])
ccat_10_dfse

array([0, 4, 3, 8, 4, 9, 1, 6, 8, 4, 0, 8, 6, 5, 9, 8, 4, 6, 4, 5, 2, 6,
       0, 5, 7, 2, 9, 7, 5, 9, 9, 5, 0, 4, 9, 5, 0, 3, 4, 1, 2, 4, 6, 5,
       4, 7, 9, 0, 0, 9, 8, 1, 1, 9, 7, 8, 0, 9, 3, 4, 1, 5, 5, 9, 8, 6,
       8, 5, 8, 7, 3, 5, 8, 4, 6, 1, 0, 0, 9, 1, 4, 8, 3, 0, 5, 4, 0, 4,
       9, 4, 7, 4, 5, 9, 6, 2, 5, 6, 3, 0, 6, 2, 8, 1, 2, 9, 7, 7, 3, 0,
       1, 1, 3, 2, 4, 8, 6, 6, 9, 5, 9, 0, 0, 7, 1, 4, 6, 2, 7, 2, 4, 1,
       4, 1, 9, 1, 9, 2, 1, 8, 3, 6, 9, 0, 2, 8, 7, 2, 2, 0, 8, 7, 5, 7,
       9, 8, 1, 7, 9, 5, 6, 6, 9, 7, 1, 7, 9, 8, 1, 8, 6, 5, 2, 6, 6, 5,
       4, 1, 0, 8, 3, 4, 7, 4, 0, 7, 7, 5, 9, 2, 2, 0, 0, 9, 2, 1, 7, 4,
       6, 3])

In [37]:
ccat_10_dfse_true = "1, 4, 3, 8, 4, 9, 1, 6, 8, 4, 1, 3, 6, 5, 3, 8, 4, 6, 4, 5, 2, 6, 0, 5, 2, 2, 9, 7, 5, 9, 9, 5, 0, 4, 8, 5, 0, 3, 4, 1, 7, 2, 6, 5, 7, 2, 9, 0, 0, 9, 8, 1, 1, 3, 1, 8, 0, 9, 3, 2, 1, 5, 5, 3, 8, 6, 8, 5, 9, 7, 3, 5, 8, 4, 6, 1, 0, 0, 9, 1, 4, 8, 3, 0, 5, 4, 0, 4, 9, 4, 7, 7, 5, 9, 6, 2, 5, 6, 3, 0, 6, 2, 8, 3, 2, 9, 2, 7, 3, 0, 1, 1, 3, 2, 4, 8, 6, 6, 9, 5, 9, 0, 0, 7, 1, 4, 6, 2, 7, 2, 4, 1, 2, 8, 3, 1, 9, 2, 1, 8, 3, 6, 3, 0, 7, 8, 7, 2, 2, 0, 8, 2, 5, 2, 9, 8, 1, 7, 9, 5, 6, 6, 3, 7, 1, 7, 9, 8, 1, 8, 6, 5, 4, 6, 6, 5, 7, 1, 0, 8, 3, 4, 7, 7, 0, 7, 7, 5, 9, 2, 4, 0, 0, 9, 4, 3, 7, 4, 6, 3"
ccat_10_dfse_true = np.array([int(i) for i in ccat_10_dfse_true.split(',')])

In [38]:
ccat_10_dfse_linear_svc = "5, 0, 7, 1, 4, 0, 6, 6, 0, 7, 2, 7, 9, 1, 0, 7, 8, 5, 6, 8, 9, 2, 3, 9, 0, 3, 1, 4, 2, 0, 1, 7, 0, 8, 6, 5, 0, 3, 8, 9, 2, 9, 8, 5, 7, 1, 1, 6, 5, 3, 6, 2, 2, 5, 3, 1, 8, 9, 1, 5, 9, 9, 2, 7, 3, 4, 7, 0, 6, 6, 2, 6, 9, 0, 2, 4, 8, 0, 9, 4, 1, 9, 6, 8, 0, 6, 6, 9, 4, 7, 4, 5, 8, 0, 3, 9, 2, 8, 7, 1, 2, 3, 6, 6, 7, 6, 9, 4, 0, 1, 1, 5, 6, 9, 0, 5, 1, 4, 4, 9, 8, 2, 7, 6, 5, 0, 7, 8, 0, 4, 8, 1, 0, 9, 8, 7, 4, 0, 3, 5, 5, 8, 8, 8, 9, 1, 4, 5, 8, 9, 4, 4, 3, 5, 7, 3, 1, 5, 4, 2, 4, 0, 4, 6, 1, 7, 8, 4, 7, 4, 8, 4, 6, 3, 4, 5, 4, 3, 2, 8, 6, 5, 4, 4, 5, 1, 3, 0, 9, 9, 9, 9, 7, 4, 0, 5, 7, 4, 2, 1"
ccat_10_dfse_linear_svc = np.array([int(i) for i in ccat_10_dfse_linear_svc.split(',')])

In [39]:
ccat_10_dfse_linear_svc_true = "5, 0, 2, 1, 7, 0, 6, 6, 0, 2, 2, 7, 9, 1, 0, 7, 8, 5, 6, 1, 9, 2, 3, 9, 0, 3, 1, 4, 2, 0, 1, 7, 0, 8, 6, 5, 0, 3, 3, 9, 2, 9, 8, 5, 7, 1, 1, 6, 5, 9, 6, 4, 2, 5, 3, 3, 8, 9, 1, 5, 9, 3, 2, 7, 3, 4, 7, 0, 6, 6, 2, 6, 9, 1, 4, 2, 8, 0, 9, 4, 1, 8, 6, 8, 0, 6, 6, 9, 4, 7, 4, 5, 8, 0, 3, 9, 7, 8, 7, 1, 2, 3, 6, 6, 7, 6, 9, 2, 0, 3, 1, 5, 6, 9, 0, 5, 1, 2, 7, 9, 8, 4, 2, 6, 5, 0, 7, 8, 0, 2, 8, 8, 0, 3, 8, 2, 4, 0, 3, 5, 5, 8, 8, 8, 3, 1, 7, 5, 8, 3, 4, 4, 3, 5, 7, 3, 1, 5, 4, 7, 4, 0, 2, 6, 1, 7, 8, 4, 1, 7, 8, 7, 6, 3, 4, 5, 4, 3, 2, 9, 6, 5, 4, 4, 5, 1, 3, 0, 9, 9, 9, 9, 7, 4, 1, 5, 2, 4, 2, 1"
ccat_10_dfse_linear_svc_true = np.array([int(i) for i in ccat_10_dfse_linear_svc_true.split(',')])

In [40]:
a_ok, b_ok = (ccat_10_dfse == ccat_10_dfse_true), (ccat_10_dfse_linear_svc == ccat_10_dfse_linear_svc_true)
table = [[np.sum(a_ok & b_ok),  np.sum(a_ok & ~b_ok)],
         [np.sum(~a_ok & b_ok), np.sum(~a_ok & ~b_ok)]]
print(mcnemar(table, exact=True))   # two-sided p; halve it if pb gains over pa

pvalue      0.794843653982861
statistic   28.0


In [41]:
print("both right:", np.sum(a_ok & b_ok))
print("only pa right:", np.sum(a_ok & ~b_ok))   # should be 46 if pa = ensemble
print("only pb right:", np.sum(~a_ok & b_ok))   # should be 21 if pb = Linear SVC
print("acc pa:", a_ok.mean(), "acc pb:", b_ok.mean())

both right: 137
only pa right: 31
only pb right: 28
acc pa: 0.84 acc pb: 0.825


On CCAT_10, the ensemble was ahead by 1.5 points on one 200-document fold, but the difference was not statistically significant (exact McNemar p = 0.79; 31 vs 28 discordant documents). I therefore do not claim a significant improvement on CCAT_10. With 1,000 documents, the dataset is too small to detect gains of this size

In [42]:
fanfic_22_dfse = '''17 15 16 12 17  3  6 13 18 18 10 10 13 15  5 14 14 15 10  7 10 10 14 10
 10 16 10  2 17 10  3  3  8 13 15  8 10 12  5 14  4 10 17  8 12 10 15 16
 10  4 10  0 10 15 10  2 10 15 20 15 15 15  7  4  6  8  5 15 12 13 10 16
  5 10 16  2  3 15 16 10 10 15 10 10 10 16 16 10'''.replace("\n","")
fanfic_22_dfse = np.array([int(i) for i in fanfic_22_dfse.split()])

In [43]:
fanfic_22_dfse_true = '''17  9 16 12 17  3  6 13 18 18 20 11 13  4  5 14 14 19  7 18 15  7 14  6
  0 19 11  2 17  6  3  3  8 13 19  8  0 12  5 14  4 21 17  8 12 11 15 16
 10  4 10  0 18 15 21  2  7  1 20 19  1  9 21  4  6  8  5  1 12 13 11  2
  5 20 16  2  3  1 15  0 10  9 10  7 21 16  9 20'''.replace("\n","")
fanfic_22_dfse_true = np.array([int(i) for i in fanfic_22_dfse_true.split()])

In [45]:
fanfic_22_linear_svc = '''17 15 16 12 17  3  6 13 18 18 10 10 13 15  5 14 14 15 10  6 10 10 14  1
 10 16 10  2 17 15  3  3  8 13  4  8 10 12  5 14  4 10 17  8 12 10 15 15
 10  4 10  0 10 15 10  1 10 15 10 15 15  9  7  4  6  8  5  1 12 13 10 16
  5 10 16  1  3 15 16  0 10 15 10  0 10 15  1 10'''.replace("\n","")
fanfic_22_linear_svc = np.array([int(i) for i in fanfic_22_linear_svc.split()])

In [46]:
fanfic_22_linear_svc_true = '''17  9 16 12 17  3  6 13 18 18 20 11 13  4  5 14 14 19  7 18 15  7 14  6
  0 19 11  2 17  6  3  3  8 13 19  8  0 12  5 14  4 21 17  8 12 11 15 16
 10  4 10  0 18 15 21  2  7  1 20 19  1  9 21  4  6  8  5  1 12 13 11  2
  5 20 16  2  3  1 15  0 10  9 10  7 21 16  9 20'''
fanfic_22_linear_svc_true = np.array([int(i) for i in fanfic_22_linear_svc_true.split()])

In [47]:
a_ok, b_ok = (fanfic_22_dfse == fanfic_22_dfse_true), (fanfic_22_linear_svc == fanfic_22_linear_svc_true)
table = [[np.sum(a_ok & b_ok),  np.sum(a_ok & ~b_ok)],
         [np.sum(~a_ok & b_ok), np.sum(~a_ok & ~b_ok)]]
print(mcnemar(table, exact=True))   # two-sided p; halve it if pb gains over pa

pvalue      0.7265625
statistic   3.0


In [48]:
print("both right:", np.sum(a_ok & b_ok))
print("only pa right:", np.sum(a_ok & ~b_ok))   # should be 46 if pa = ensemble
print("only pb right:", np.sum(~a_ok & b_ok))   # should be 21 if pb = Linear SVC
print("acc pa:", a_ok.mean(), "acc pb:", b_ok.mean())

both right: 45
only pa right: 5
only pb right: 3
acc pa: 0.5681818181818182 acc pb: 0.5454545454545454


On Fanfic_22, the ensemble was nominally ahead on one 88-document fold (56.8% vs 54.5%), but the models disagreed on only 8 documents, and the difference was not statistically significant (exact McNemar p = 0.73). With 440 documents in total, the dataset is too small to detect gains of about 1 to 2 points, so I report the direction of the improvement and do not claim significance on this dataset.

In [49]:
imdb_62_dfse = "53, 28, 51, 59, 46, 38, 16, 55, 30, 50, 7, 30, 28, 53, 45, 39, 54, 6, 0, 44, 10, 26, 39, 20, 10, 27, 42, 33, 22, 47, 25, 38, 40, 0, 20, 20, 7, 10, 47, 55, 5, 50, 3, 53, 24, 50, 25, 32, 29, 38, 11, 15, 30, 9, 38, 58, 60, 54, 20, 35, 16, 11, 9, 47, 0, 2, 32, 54, 34, 4, 53, 0, 27, 39, 11, 2, 25, 42, 60, 55, 21, 41, 42, 25, 56, 49, 41, 47, 32, 1, 21, 52, 42, 45, 3, 5, 50, 43, 7, 26, 30, 43, 33, 40, 26, 25, 13, 26, 9, 14, 5, 3, 50, 53, 60, 54, 3, 10, 52, 55, 34, 55, 25, 5, 53, 14, 9, 1, 23, 15, 48, 49, 57, 46, 11, 53, 36, 52, 43, 0, 43, 14, 48, 43, 5, 9, 38, 34, 25, 38, 59, 52, 17, 4, 36, 54, 56, 52, 10, 34, 13, 35, 23, 18, 38, 0, 55, 57, 26, 30, 57, 13, 41, 39, 27, 59, 55, 36, 45, 54, 34, 44, 20, 0, 32, 10, 51, 10, 44, 4, 5, 46, 11, 21, 57, 44, 9, 27, 25, 33, 13, 39, 21, 15, 30, 19, 23, 24, 29, 57, 41, 16, 23, 60, 53, 1, 51, 26, 18, 24, 33, 33, 33, 5, 41, 41, 37, 9, 15, 38, 61, 40, 1, 35, 26, 6, 9, 33, 26, 56, 25, 11, 60, 21, 34, 44, 59, 52, 38, 20, 32, 33, 61, 5, 15, 60, 23, 2, 1, 57, 9, 53, 1, 7, 25, 25, 60, 26, 5, 46, 39, 18, 20, 52, 33, 7, 18, 39, 36, 11, 57, 10, 17, 12, 4, 53, 39, 1, 18, 20, 8, 29, 40, 41, 18, 2, 0, 18, 33, 41, 22, 0, 11, 19, 39, 1, 26, 16, 34, 15, 22, 25, 50, 6, 46, 35, 19, 4, 12, 10, 56, 11, 6, 33, 59, 22, 24, 1, 6, 49, 23, 2, 15, 58, 38, 7, 0, 5, 8, 6, 39, 0, 33, 39, 11, 43, 0, 32, 4, 10, 49, 26, 39, 1, 44, 2, 18, 34, 31, 35, 8, 16, 7, 8, 39, 3, 7, 5, 51, 30, 0, 53, 56, 5, 31, 12, 53, 10, 6, 53, 5, 45, 58, 0, 20, 25, 19, 21, 43, 55, 5, 24, 38, 15, 21, 15, 34, 59, 15, 36, 1, 7, 34, 19, 2, 16, 37, 38, 58, 16, 9, 24, 1, 39, 32, 2, 35, 60, 6, 10, 2, 10, 25, 44, 53, 30, 51, 11, 11, 18, 32, 6, 43, 0, 40, 30, 50, 48, 13, 4, 39, 43, 60, 39, 31, 50, 27, 42, 33, 47, 11, 41, 32, 12, 3, 39, 10, 25, 36, 9, 16, 17, 36, 54, 33, 11, 31, 43, 30, 46, 46, 50, 53, 53, 30, 59, 47, 16, 32, 40, 5, 18, 0, 53, 25, 43, 19, 61, 27, 9, 6, 29, 14, 7, 25, 38, 58, 41, 34, 12, 61, 32, 9, 11, 16, 27, 2, 35, 5, 35, 46, 11, 43, 55, 9, 56, 39, 4, 12, 50, 47, 3, 61, 30, 60, 22, 6, 18, 11, 9, 21, 0, 20, 2, 38, 49, 4, 38, 16, 2, 53, 10, 32, 5, 24, 2, 16, 11, 43, 0, 4, 5, 7, 36, 33, 17, 60, 38, 5, 42, 61, 7, 50, 54, 49, 7, 39, 28, 13, 42, 34, 61, 39, 22, 40, 15, 37, 4, 37, 21, 35, 3, 24, 40, 9, 9, 9, 50, 28, 35, 15, 0, 4, 15, 21, 43, 39, 14, 6, 34, 34, 10, 57, 59, 33, 39, 18, 10, 33, 33, 46, 21, 55, 39, 42, 3, 24, 33, 20, 2, 20, 37, 16, 34, 28, 10, 54, 59, 49, 38, 27, 55, 41, 37, 26, 23, 39, 54, 19, 35, 6, 31, 1, 37, 16, 0, 2, 54, 35, 33, 5, 61, 12, 52, 14, 41, 58, 28, 36, 36, 17, 10, 2, 45, 42, 3, 40, 61, 44, 5, 41, 37, 57, 7, 58, 15, 47, 10, 54, 41, 9, 47, 34, 36, 35, 7, 41, 43, 49, 5, 36, 40, 44, 8, 33, 10, 51, 57, 24, 48, 2, 24, 36, 44, 42, 5, 49, 6, 50, 38, 45, 36, 51, 41, 11, 40, 50, 10, 34, 20, 4, 61, 38, 3, 9, 18, 16, 54, 14, 22, 25, 6, 51, 36, 16, 50, 8, 44, 1, 21, 50, 0, 27, 36, 17, 2, 9, 55, 42, 24, 57, 50, 36, 49, 8, 5, 13, 5, 16, 33, 50, 61, 1, 35, 25, 9, 31, 51, 33, 10, 35, 9, 41, 44, 5, 1, 39, 5, 22, 20, 10, 37, 14, 53, 37, 30, 44, 2, 3, 58, 16, 1, 50, 5, 41, 21, 19, 8, 11, 58, 24, 0, 40, 9, 20, 52, 5, 17, 15, 44, 17, 56, 39, 16, 8, 44, 20, 4, 30, 33, 48, 7, 11, 19, 37, 2, 53, 22, 32, 53, 25, 27, 5, 15, 13, 45, 55, 9, 40, 34, 5, 27, 1, 10, 50, 41, 52, 12, 51, 47, 61, 44, 4, 4, 37, 57, 33, 25, 25, 1, 41, 50, 37, 25, 25, 21, 34, 45, 5, 23, 38, 49, 2, 18, 1, 11, 58, 44, 48, 52, 21, 29, 11, 32, 26, 14, 41, 22, 35, 7, 35, 35, 37, 36, 9, 26, 11, 34, 13, 7, 43, 56, 13, 61, 17, 2, 33, 5, 6, 22, 49, 11, 54, 4, 57, 30, 58, 1, 37, 16, 23, 45, 29, 55, 50, 10, 36, 5, 19, 2, 10, 7, 35, 5, 2, 18, 7, 40, 35, 2, 27, 58, 36, 55, 30, 60, 33, 48, 5, 36, 0, 39, 61, 42, 21, 12, 3, 48, 7, 45, 12, 1, 39, 1, 43, 50, 31, 17, 39, 31, 45, 36, 28, 31, 34, 24, 49, 57, 50, 1, 16, 10, 60, 44, 27, 10, 50, 28, 50, 30, 30, 27, 54, 19, 48, 1, 11, 42, 49, 39, 42, 28, 49, 26, 57, 54, 56, 53, 44, 14, 43, 55, 10, 36, 58, 51, 59, 61, 33, 9, 39, 13, 52, 51, 40, 37, 1, 27, 24, 16, 25, 28, 37, 20, 25, 30, 46, 15, 14, 9, 18, 6, 61, 3, 13, 5, 14, 20, 15, 9, 6, 34, 36, 2, 46, 9, 38, 7, 40, 22, 50, 38, 3, 5, 55, 45, 7, 32, 43, 53, 2, 26, 54, 7, 5, 36, 53, 41, 10, 53, 47, 45, 15, 36, 55, 1, 34, 40, 54, 21, 2, 12, 13, 52, 58, 51, 60, 36, 58, 4, 32, 59, 24, 15, 57, 10, 42, 34, 55, 15, 35, 27, 28, 4, 1, 60, 58, 5, 58, 38, 13, 38, 46, 50, 57, 53, 12, 20, 5, 52, 24, 28, 35, 38, 53, 60, 15, 30, 1, 13, 5, 33, 41, 15, 30, 21, 30, 39, 4, 1, 53, 5, 11, 14, 33, 7, 40, 57, 16, 2, 33, 1, 25, 36, 39, 15, 12, 14, 19, 1, 41, 12, 58, 58, 48, 52, 47, 1, 30, 16, 28, 33, 49, 53, 11, 44, 59, 8, 40, 5, 9, 18, 39, 57, 60, 13, 52, 30, 54, 18, 24, 58, 25, 5, 28, 54, 17, 18, 13, 14, 36, 5, 15, 20, 22, 53, 43, 54, 45, 49, 60, 52, 52, 11, 2, 33, 59, 26, 48, 5, 26, 29, 59, 45, 9, 0, 22, 36, 50, 40"
imdb_62_dfse = np.array([int(i) for i in imdb_62_dfse.split(',')])

In [50]:
imdb_62_dfse_true = "0, 27, 51, 59, 2, 38, 44, 33, 30, 50, 31, 30, 27, 47, 45, 44, 41, 6, 36, 44, 10, 26, 59, 20, 10, 27, 24, 3, 22, 7, 25, 13, 40, 48, 20, 20, 46, 55, 46, 53, 29, 12, 3, 60, 24, 50, 25, 32, 29, 38, 11, 15, 30, 33, 38, 58, 60, 54, 25, 35, 58, 33, 9, 46, 14, 39, 32, 58, 34, 4, 0, 50, 27, 7, 11, 16, 60, 24, 10, 55, 21, 41, 19, 1, 56, 49, 0, 2, 32, 52, 21, 52, 49, 45, 3, 42, 24, 43, 51, 26, 12, 43, 42, 40, 26, 25, 13, 26, 48, 22, 61, 3, 44, 23, 60, 5, 3, 10, 52, 55, 27, 55, 25, 59, 53, 14, 12, 23, 30, 15, 48, 49, 57, 47, 11, 47, 61, 52, 43, 36, 43, 41, 48, 43, 50, 54, 31, 34, 23, 23, 59, 52, 17, 4, 36, 50, 56, 53, 6, 34, 13, 35, 30, 18, 41, 0, 9, 57, 26, 51, 57, 13, 50, 39, 8, 59, 55, 0, 45, 48, 27, 44, 20, 36, 32, 17, 38, 48, 56, 34, 58, 46, 11, 21, 57, 39, 4, 27, 47, 48, 13, 39, 36, 15, 35, 42, 2, 24, 29, 57, 29, 16, 2, 60, 39, 21, 51, 26, 18, 33, 29, 6, 33, 5, 0, 2, 34, 31, 27, 54, 61, 40, 10, 35, 26, 17, 37, 19, 26, 56, 55, 23, 60, 21, 34, 32, 59, 52, 21, 47, 32, 38, 61, 59, 15, 20, 38, 2, 7, 57, 9, 12, 12, 23, 46, 55, 60, 26, 42, 46, 59, 18, 25, 52, 31, 31, 18, 28, 49, 11, 57, 10, 46, 12, 23, 46, 44, 10, 18, 20, 8, 41, 40, 39, 18, 7, 39, 18, 33, 41, 22, 48, 53, 19, 44, 13, 26, 23, 34, 15, 22, 25, 50, 15, 46, 35, 28, 32, 26, 10, 56, 8, 60, 14, 59, 22, 24, 1, 34, 49, 31, 2, 15, 58, 30, 48, 0, 41, 8, 2, 0, 0, 39, 39, 51, 43, 60, 32, 53, 53, 49, 26, 16, 6, 19, 12, 18, 34, 48, 35, 8, 19, 20, 23, 8, 3, 7, 42, 47, 30, 41, 20, 56, 42, 2, 23, 38, 2, 6, 38, 44, 45, 58, 29, 20, 47, 42, 21, 43, 55, 5, 24, 53, 12, 21, 42, 34, 59, 1, 47, 21, 25, 34, 19, 15, 16, 30, 53, 5, 54, 9, 24, 1, 39, 32, 2, 35, 60, 6, 1, 1, 10, 47, 44, 46, 30, 51, 51, 11, 18, 32, 15, 43, 53, 40, 38, 50, 54, 13, 44, 59, 43, 47, 56, 22, 50, 27, 56, 11, 31, 23, 41, 32, 23, 3, 39, 11, 7, 8, 9, 58, 17, 0, 33, 50, 39, 2, 43, 2, 37, 46, 16, 53, 11, 30, 46, 47, 44, 32, 40, 29, 18, 55, 7, 25, 43, 19, 61, 8, 9, 6, 29, 14, 10, 25, 38, 4, 0, 27, 12, 61, 32, 54, 11, 16, 27, 1, 35, 5, 35, 46, 54, 43, 53, 4, 56, 53, 16, 12, 45, 12, 3, 61, 30, 54, 22, 6, 18, 11, 9, 21, 41, 6, 4, 7, 49, 31, 12, 16, 22, 31, 23, 32, 8, 50, 38, 16, 6, 43, 14, 4, 29, 51, 53, 42, 17, 6, 46, 54, 54, 61, 54, 46, 6, 49, 7, 39, 28, 13, 42, 34, 61, 59, 22, 40, 15, 37, 28, 37, 21, 35, 3, 24, 40, 4, 16, 4, 50, 28, 35, 22, 1, 32, 17, 21, 43, 39, 14, 13, 34, 17, 38, 57, 42, 29, 24, 18, 1, 33, 33, 51, 21, 14, 59, 56, 3, 24, 38, 20, 2, 20, 37, 48, 17, 28, 31, 5, 33, 49, 38, 30, 54, 14, 37, 26, 23, 2, 28, 19, 37, 15, 31, 36, 37, 50, 0, 54, 6, 35, 33, 42, 61, 51, 52, 14, 36, 58, 28, 29, 18, 17, 47, 2, 45, 56, 3, 40, 61, 52, 50, 15, 37, 57, 55, 26, 19, 36, 10, 5, 8, 23, 44, 34, 36, 35, 10, 41, 43, 49, 8, 36, 40, 45, 8, 54, 10, 12, 57, 29, 48, 9, 24, 0, 44, 32, 41, 49, 22, 45, 12, 45, 36, 51, 7, 7, 40, 39, 1, 57, 23, 58, 61, 20, 3, 9, 18, 19, 38, 14, 22, 25, 6, 5, 36, 24, 4, 55, 39, 48, 21, 19, 41, 27, 36, 17, 51, 9, 5, 24, 8, 57, 39, 5, 49, 22, 11, 13, 33, 24, 9, 50, 61, 33, 37, 47, 37, 48, 51, 42, 10, 35, 9, 8, 9, 41, 31, 59, 29, 22, 10, 10, 37, 14, 0, 44, 30, 5, 56, 3, 44, 16, 1, 29, 56, 41, 21, 19, 8, 11, 58, 24, 41, 40, 49, 20, 52, 0, 17, 1, 16, 17, 56, 29, 16, 8, 44, 20, 4, 51, 8, 48, 38, 11, 19, 37, 4, 38, 3, 32, 55, 25, 7, 42, 15, 13, 45, 22, 9, 40, 4, 5, 27, 11, 60, 5, 42, 52, 7, 31, 60, 61, 28, 4, 4, 37, 34, 15, 48, 20, 0, 41, 50, 37, 55, 25, 21, 27, 45, 29, 23, 38, 49, 7, 18, 1, 2, 58, 58, 4, 52, 21, 51, 8, 32, 26, 14, 41, 24, 35, 46, 35, 35, 37, 36, 12, 26, 58, 34, 59, 51, 43, 56, 13, 61, 17, 7, 24, 19, 6, 22, 49, 17, 12, 9, 57, 30, 16, 6, 37, 16, 48, 45, 29, 55, 3, 31, 14, 5, 28, 31, 10, 0, 35, 28, 46, 18, 7, 40, 35, 7, 28, 45, 0, 59, 27, 60, 5, 16, 19, 8, 33, 33, 61, 56, 21, 12, 3, 48, 25, 45, 12, 55, 42, 1, 43, 28, 0, 17, 56, 31, 45, 55, 28, 2, 34, 44, 49, 57, 39, 23, 16, 36, 60, 27, 17, 38, 50, 44, 53, 30, 31, 27, 7, 19, 30, 1, 11, 33, 49, 50, 42, 28, 49, 26, 57, 15, 56, 46, 58, 14, 43, 55, 10, 36, 5, 51, 19, 61, 33, 12, 33, 13, 52, 54, 40, 37, 1, 27, 24, 58, 25, 28, 26, 60, 23, 11, 47, 22, 14, 9, 18, 6, 61, 3, 13, 5, 14, 20, 36, 38, 48, 34, 46, 16, 46, 37, 52, 54, 40, 22, 50, 30, 3, 5, 55, 45, 6, 32, 43, 25, 53, 34, 51, 31, 3, 36, 53, 41, 20, 22, 31, 45, 15, 1, 55, 1, 34, 40, 58, 21, 59, 4, 13, 52, 58, 47, 7, 14, 58, 4, 32, 13, 61, 1, 57, 10, 50, 2, 55, 54, 35, 8, 42, 28, 6, 60, 58, 39, 27, 47, 13, 25, 47, 25, 57, 23, 12, 20, 11, 52, 56, 28, 37, 51, 53, 60, 15, 53, 10, 13, 61, 3, 5, 15, 30, 21, 54, 24, 9, 36, 56, 19, 11, 14, 42, 47, 40, 57, 16, 33, 29, 6, 20, 36, 44, 1, 47, 14, 27, 15, 53, 9, 45, 58, 25, 52, 47, 31, 30, 7, 28, 33, 49, 60, 11, 51, 59, 17, 40, 29, 9, 18, 39, 57, 60, 13, 52, 30, 4, 18, 42, 44, 25, 54, 28, 9, 17, 18, 13, 14, 17, 19, 15, 20, 22, 53, 43, 5, 45, 49, 60, 52, 52, 17, 2, 19, 59, 26, 48, 29, 26, 41, 59, 45, 4, 0, 57, 54, 16, 40"
imdb_62_dfse_true = np.array([int(i) for i in imdb_62_dfse_true.split(',')])

In [51]:
imdb_62_linear_svc = '46, 43, 15, 1, 42, 25, 27, 33, 25, 46, 36, 9, 44, 59, 49, 12, 44, 36, 31, 23, 50, 24, 38, 49, 31, 60, 61, 11, 25, 21, 3, 49, 53, 24, 24, 38, 14, 22, 40, 9, 36, 55, 43, 18, 38, 9, 1, 31, 26, 4, 49, 23, 20, 58, 29, 40, 28, 24, 15, 54, 5, 27, 34, 2, 40, 11, 54, 19, 9, 21, 5, 38, 25, 21, 48, 2, 41, 7, 42, 2, 2, 39, 38, 41, 60, 24, 9, 24, 35, 28, 30, 54, 9, 20, 38, 51, 12, 43, 10, 21, 30, 8, 56, 57, 0, 50, 16, 41, 37, 52, 39, 20, 14, 31, 5, 55, 53, 20, 17, 60, 11, 42, 58, 38, 15, 60, 33, 10, 2, 32, 10, 36, 11, 11, 15, 1, 8, 1, 34, 24, 23, 2, 41, 8, 24, 12, 5, 28, 51, 26, 48, 4, 45, 18, 21, 16, 49, 7, 39, 6, 25, 53, 1, 12, 8, 55, 4, 34, 21, 35, 4, 23, 45, 1, 42, 53, 45, 50, 53, 41, 45, 5, 14, 38, 1, 15, 51, 10, 18, 9, 1, 26, 53, 39, 49, 12, 5, 16, 53, 17, 49, 51, 50, 16, 15, 49, 37, 61, 15, 25, 34, 28, 27, 6, 35, 33, 57, 35, 2, 12, 33, 9, 28, 20, 30, 33, 31, 11, 54, 45, 20, 30, 22, 53, 57, 20, 29, 29, 22, 34, 0, 1, 26, 0, 31, 20, 28, 43, 24, 16, 2, 31, 46, 52, 52, 26, 44, 60, 2, 15, 33, 34, 54, 57, 13, 11, 28, 54, 11, 49, 4, 61, 10, 49, 61, 55, 44, 53, 23, 17, 8, 38, 10, 54, 12, 47, 15, 44, 1, 44, 13, 5, 9, 5, 5, 11, 26, 21, 15, 55, 55, 12, 46, 8, 54, 19, 53, 18, 30, 30, 11, 2, 5, 32, 31, 50, 61, 14, 18, 58, 10, 9, 16, 40, 5, 32, 58, 21, 30, 26, 61, 10, 60, 34, 35, 36, 35, 17, 10, 59, 13, 26, 61, 29, 45, 0, 52, 8, 20, 21, 17, 58, 15, 32, 58, 15, 7, 43, 2, 27, 60, 31, 61, 36, 30, 36, 22, 38, 7, 55, 38, 38, 8, 37, 43, 11, 45, 38, 33, 35, 16, 22, 11, 11, 33, 61, 28, 35, 12, 55, 55, 48, 13, 12, 41, 0, 53, 46, 1, 7, 61, 26, 26, 33, 0, 39, 41, 53, 41, 53, 36, 35, 14, 45, 22, 43, 0, 16, 26, 9, 35, 18, 5, 39, 54, 18, 35, 15, 59, 33, 37, 32, 34, 59, 11, 6, 12, 11, 50, 25, 27, 15, 31, 13, 34, 2, 38, 19, 43, 0, 32, 50, 39, 55, 16, 24, 56, 0, 16, 11, 30, 50, 53, 0, 38, 33, 53, 8, 60, 14, 28, 40, 7, 41, 60, 25, 9, 28, 9, 11, 47, 60, 46, 41, 40, 19, 55, 39, 27, 13, 58, 38, 31, 20, 39, 51, 1, 15, 9, 38, 21, 54, 15, 32, 32, 41, 55, 6, 1, 53, 3, 39, 38, 39, 58, 35, 11, 19, 9, 22, 54, 17, 1, 3, 28, 40, 39, 22, 57, 25, 57, 18, 9, 20, 25, 24, 4, 39, 2, 38, 44, 39, 6, 38, 60, 35, 1, 15, 57, 34, 20, 36, 12, 61, 35, 44, 14, 33, 11, 16, 53, 57, 36, 28, 6, 40, 4, 2, 5, 59, 21, 5, 38, 15, 15, 2, 32, 6, 49, 5, 6, 9, 59, 36, 50, 7, 36, 22, 43, 1, 40, 30, 53, 53, 42, 20, 56, 52, 1, 36, 26, 18, 1, 14, 57, 2, 16, 21, 20, 51, 9, 60, 30, 59, 54, 30, 24, 0, 14, 24, 34, 11, 55, 2, 13, 43, 54, 11, 28, 1, 34, 7, 38, 59, 53, 41, 12, 3, 50, 61, 7, 58, 38, 33, 24, 58, 10, 40, 39, 7, 55, 42, 26, 13, 61, 53, 0, 9, 16, 22, 9, 21, 30, 27, 10, 3, 40, 50, 39, 51, 43, 35, 0, 39, 10, 57, 45, 57, 31, 34, 20, 43, 41, 44, 29, 51, 8, 13, 49, 53, 61, 43, 53, 21, 19, 39, 55, 42, 13, 7, 32, 39, 46, 25, 37, 15, 5, 10, 30, 42, 20, 34, 39, 18, 38, 38, 15, 31, 25, 36, 53, 46, 33, 56, 19, 27, 56, 42, 53, 58, 42, 38, 37, 18, 35, 45, 16, 0, 21, 59, 3, 32, 0, 6, 25, 50, 38, 52, 32, 18, 26, 59, 26, 31, 33, 7, 19, 7, 6, 50, 25, 38, 22, 49, 27, 11, 15, 37, 4, 5, 41, 14, 27, 35, 38, 46, 11, 59, 45, 38, 9, 10, 39, 39, 45, 52, 40, 52, 0, 22, 10, 41, 13, 11, 7, 39, 15, 51, 38, 18, 37, 2, 53, 18, 29, 30, 31, 42, 53, 41, 12, 48, 0, 58, 21, 7, 58, 11, 11, 2, 37, 26, 52, 13, 50, 54, 7, 33, 35, 18, 19, 20, 38, 55, 39, 0, 39, 57, 41, 55, 0, 36, 6, 6, 22, 39, 34, 35, 21, 2, 17, 2, 11, 14, 42, 24, 32, 23, 26, 43, 26, 40, 16, 28, 7, 15, 16, 0, 59, 51, 46, 52, 37, 20, 25, 2, 44, 4, 4, 34, 4, 13, 20, 46, 22, 46, 41, 48, 58, 37, 10, 50, 41, 22, 25, 24, 53, 9, 52, 29, 35, 15, 10, 37, 34, 1, 15, 34, 51, 8, 2, 41, 32, 24, 18, 24, 55, 36, 0, 23, 31, 33, 38, 18, 25, 28, 41, 9, 20, 32, 3, 2, 39, 22, 34, 0, 2, 10, 17, 35, 44, 5, 15, 17, 2, 0, 30, 34, 36, 50, 13, 54, 55, 8, 37, 11, 27, 43, 2, 39, 42, 2, 7, 6, 26, 9, 13, 61, 42, 23, 42, 59, 48, 21, 11, 1, 11, 43, 2, 52, 3, 38, 45, 1, 52, 31, 38, 7, 4, 3, 13, 44, 56, 13, 0, 3, 2, 4, 50, 25, 18, 7, 9, 45, 20, 9, 1, 53, 1, 10, 57, 37, 11, 22, 51, 2, 29, 60, 4, 49, 43, 2, 50, 21, 47, 18, 16, 11, 61, 57, 16, 58, 58, 8, 3, 1, 39, 36, 24, 36, 9, 60, 23, 61, 3, 49, 9, 60, 49, 41, 10, 37, 40, 53, 61, 53, 3, 15, 57, 41, 38, 9, 27, 31, 38, 25, 3, 6, 49, 10, 38, 40, 53, 2, 7, 4, 6, 4, 39, 34, 36, 40, 24, 52, 48, 27, 21, 10, 7, 30, 59, 53, 26, 39, 0, 8, 2, 33, 21, 16, 10, 2, 5, 44, 47, 58, 56, 28, 6, 40, 54, 24, 54, 2, 41, 31, 54, 7, 32, 55, 41, 60, 2, 45, 25, 0, 55, 37, 37, 40, 44, 54, 43, 26, 25, 37, 42, 6, 56, 21, 39, 61, 20, 14, 53, 2, 17, 47, 52, 13, 36, 57, 17, 25, 37, 50, 9, 11, 19, 0, 21, 40, 5, 3, 60, 22, 42, 23, 37, 11, 16, 22, 58, 39, 24, 50, 43, 56, 12, 55, 51, 57, 0, 33, 51, 22, 4, 55, 12, 48, 46, 25, 33, 53, 9, 32, 41, 0, 50, 27, 5, 2, 28, 9, 31, 11, 58, 55, 42, 13, 23, 0, 1, 39, 52, 11, 15, 22, 33, 38, 40, 52, 0, 18, 5, 10, 57, 36, 43, 38, 38, 1, 24, 61, 5, 8, 36, 42, 1, 58'
imdb_62_linear_svc = np.array([int(i) for i in imdb_62_linear_svc.split(',')])

In [52]:
imdb_62_linear_svc_true = '2, 43, 15, 1, 24, 25, 8, 33, 25, 25, 17, 58, 39, 56, 49, 12, 44, 53, 25, 46, 16, 28, 31, 49, 5, 6, 28, 58, 35, 21, 3, 49, 53, 24, 7, 38, 14, 24, 40, 9, 36, 2, 43, 18, 23, 23, 1, 22, 26, 4, 49, 38, 20, 58, 29, 40, 28, 24, 15, 15, 19, 8, 27, 2, 40, 60, 16, 27, 31, 21, 5, 2, 25, 21, 25, 46, 29, 48, 24, 10, 2, 44, 20, 14, 60, 24, 9, 24, 35, 28, 30, 39, 9, 51, 29, 54, 54, 43, 17, 1, 30, 53, 56, 57, 39, 5, 24, 0, 4, 52, 42, 20, 41, 31, 19, 55, 7, 20, 17, 60, 27, 56, 34, 53, 15, 54, 33, 1, 37, 32, 36, 36, 11, 53, 47, 10, 22, 1, 48, 54, 23, 46, 41, 27, 28, 23, 50, 28, 51, 34, 30, 23, 45, 17, 21, 16, 49, 47, 50, 6, 20, 53, 6, 12, 36, 55, 28, 27, 21, 35, 4, 30, 45, 14, 56, 29, 45, 3, 48, 41, 45, 50, 14, 30, 38, 15, 51, 47, 18, 9, 33, 26, 53, 56, 49, 12, 5, 34, 56, 17, 49, 38, 16, 16, 15, 49, 37, 61, 54, 60, 4, 31, 27, 6, 35, 28, 57, 35, 2, 12, 33, 9, 28, 20, 12, 29, 44, 46, 50, 45, 25, 2, 17, 21, 57, 20, 29, 29, 22, 34, 33, 1, 26, 53, 59, 12, 19, 43, 29, 34, 52, 0, 46, 52, 52, 26, 44, 60, 4, 11, 19, 34, 39, 57, 13, 23, 42, 41, 11, 49, 4, 61, 23, 49, 61, 55, 32, 38, 31, 17, 18, 53, 10, 19, 12, 47, 15, 45, 10, 28, 13, 19, 4, 11, 19, 7, 26, 21, 8, 33, 55, 47, 46, 8, 6, 58, 48, 18, 30, 30, 51, 8, 56, 32, 31, 41, 61, 14, 18, 39, 10, 54, 28, 40, 39, 32, 58, 21, 51, 26, 61, 10, 60, 34, 37, 1, 37, 17, 10, 59, 13, 26, 61, 29, 45, 41, 52, 8, 25, 36, 17, 58, 1, 32, 45, 22, 31, 43, 38, 7, 7, 23, 61, 0, 30, 55, 22, 23, 51, 53, 23, 5, 29, 37, 43, 51, 44, 33, 33, 35, 48, 1, 11, 11, 38, 61, 42, 37, 31, 5, 9, 48, 29, 12, 41, 36, 23, 31, 1, 48, 61, 26, 58, 3, 0, 29, 8, 25, 39, 38, 46, 35, 14, 45, 22, 43, 1, 44, 26, 12, 35, 18, 5, 16, 16, 18, 35, 15, 59, 39, 37, 32, 27, 59, 11, 47, 9, 2, 45, 47, 27, 53, 60, 13, 17, 15, 38, 28, 43, 60, 32, 50, 41, 14, 54, 27, 56, 8, 16, 6, 30, 50, 31, 55, 44, 49, 2, 8, 60, 14, 28, 40, 30, 51, 48, 23, 9, 4, 30, 11, 2, 60, 46, 61, 40, 42, 59, 56, 16, 13, 5, 33, 48, 20, 24, 34, 23, 15, 49, 38, 52, 48, 15, 32, 32, 41, 55, 6, 10, 0, 3, 33, 30, 36, 58, 35, 54, 19, 33, 23, 5, 17, 15, 3, 28, 40, 0, 22, 57, 48, 57, 18, 37, 20, 25, 33, 28, 56, 39, 11, 16, 59, 53, 12, 54, 35, 36, 15, 57, 34, 20, 36, 7, 61, 35, 39, 14, 38, 33, 58, 47, 34, 36, 44, 15, 40, 4, 46, 5, 59, 21, 5, 20, 1, 15, 42, 32, 6, 49, 50, 2, 48, 59, 0, 50, 31, 61, 22, 43, 13, 40, 38, 55, 46, 50, 47, 56, 52, 31, 47, 26, 18, 1, 14, 57, 51, 19, 21, 20, 5, 11, 60, 20, 42, 54, 51, 24, 41, 14, 42, 34, 19, 54, 54, 33, 43, 9, 11, 28, 8, 34, 31, 60, 59, 60, 36, 7, 3, 45, 19, 38, 44, 24, 12, 50, 58, 6, 40, 59, 10, 14, 32, 34, 13, 61, 48, 0, 9, 16, 22, 12, 21, 30, 57, 7, 3, 40, 39, 44, 51, 43, 35, 36, 42, 10, 57, 45, 57, 53, 46, 10, 43, 7, 58, 29, 12, 55, 13, 49, 25, 61, 43, 0, 21, 19, 0, 55, 44, 13, 7, 32, 42, 37, 25, 37, 15, 8, 1, 30, 56, 20, 34, 2, 18, 38, 10, 19, 9, 25, 7, 0, 46, 33, 56, 19, 27, 56, 32, 53, 26, 42, 12, 27, 18, 35, 45, 16, 36, 21, 13, 3, 32, 39, 6, 47, 53, 50, 52, 32, 18, 26, 59, 26, 9, 33, 54, 58, 38, 15, 46, 25, 5, 3, 49, 27, 11, 15, 37, 16, 52, 15, 14, 27, 35, 41, 51, 54, 59, 45, 38, 54, 10, 39, 59, 45, 53, 40, 52, 51, 57, 60, 41, 59, 54, 25, 48, 1, 24, 53, 18, 37, 2, 22, 18, 29, 30, 2, 24, 0, 41, 51, 16, 0, 45, 21, 23, 58, 58, 11, 6, 37, 26, 52, 13, 50, 58, 55, 5, 35, 18, 19, 46, 51, 22, 59, 0, 39, 57, 41, 8, 29, 0, 1, 48, 22, 39, 34, 35, 55, 54, 17, 42, 7, 14, 56, 8, 32, 2, 26, 43, 26, 40, 58, 42, 7, 22, 23, 59, 46, 51, 50, 52, 37, 20, 25, 12, 45, 4, 4, 34, 31, 13, 60, 51, 15, 46, 5, 48, 58, 37, 31, 50, 8, 22, 55, 5, 53, 9, 52, 41, 35, 22, 36, 37, 34, 6, 47, 27, 51, 17, 16, 29, 32, 24, 18, 24, 55, 54, 41, 2, 9, 6, 2, 18, 10, 28, 41, 9, 60, 32, 3, 6, 44, 17, 57, 50, 6, 10, 17, 35, 51, 14, 7, 17, 60, 14, 30, 34, 14, 50, 13, 51, 55, 8, 44, 11, 27, 43, 7, 44, 42, 31, 47, 17, 26, 26, 13, 61, 50, 2, 42, 19, 48, 55, 48, 12, 8, 43, 7, 52, 3, 25, 45, 36, 52, 2, 13, 20, 9, 3, 13, 44, 56, 13, 0, 3, 47, 9, 50, 55, 18, 6, 37, 39, 20, 4, 8, 25, 6, 10, 57, 54, 11, 22, 47, 38, 29, 7, 44, 49, 43, 46, 50, 21, 31, 18, 44, 11, 61, 57, 16, 16, 27, 8, 3, 21, 44, 49, 61, 14, 4, 47, 31, 61, 3, 49, 9, 47, 49, 33, 31, 37, 40, 39, 61, 0, 3, 1, 57, 42, 30, 4, 27, 4, 7, 60, 3, 6, 49, 10, 54, 40, 46, 42, 25, 58, 13, 4, 39, 27, 36, 40, 24, 52, 48, 27, 21, 20, 5, 35, 59, 38, 26, 29, 0, 8, 19, 3, 21, 16, 10, 27, 59, 44, 36, 44, 56, 28, 6, 40, 1, 29, 11, 2, 41, 24, 4, 31, 32, 55, 48, 60, 34, 45, 25, 39, 59, 37, 37, 40, 45, 12, 43, 26, 55, 12, 24, 6, 56, 21, 28, 3, 20, 14, 53, 22, 17, 46, 52, 13, 36, 57, 17, 7, 30, 4, 16, 5, 42, 42, 21, 40, 59, 3, 60, 56, 19, 23, 9, 5, 16, 22, 58, 39, 24, 50, 43, 56, 23, 52, 47, 57, 0, 33, 47, 22, 9, 55, 4, 30, 46, 47, 42, 20, 38, 32, 0, 14, 12, 34, 29, 11, 28, 53, 10, 23, 58, 33, 33, 13, 30, 36, 1, 19, 52, 11, 36, 22, 42, 12, 40, 52, 41, 18, 11, 10, 57, 5, 43, 38, 7, 1, 19, 61, 41, 8, 17, 24, 23, 4'
imdb_62_linear_svc_true = np.array([int(i) for i in imdb_62_linear_svc_true.split(',')])

In [53]:
a_ok, b_ok = (imdb_62_dfse == imdb_62_dfse_true), (imdb_62_linear_svc == imdb_62_linear_svc_true)
table = [[np.sum(a_ok & b_ok),  np.sum(a_ok & ~b_ok)],
         [np.sum(~a_ok & b_ok), np.sum(~a_ok & ~b_ok)]]
print(mcnemar(table, exact=True))   # two-sided p; halve it if pb gains over pa

pvalue      0.19174497919169717
statistic   284.0


In [54]:
print("both right:", np.sum(a_ok & b_ok))
print("only pa right:", np.sum(a_ok & ~b_ok))   # should be 46 if pa = ensemble
print("only pb right:", np.sum(~a_ok & b_ok))   # should be 21 if pb = Linear SVC
print("acc pa:", a_ok.mean(), "acc pb:", b_ok.mean())

both right: 317
only pa right: 317
only pb right: 284
acc pa: 0.5112903225806451 acc pb: 0.4846774193548387


On Imdb_62, the ensemble was ahead by 2.7 points on one 1,240-document fold, but the difference was not significant (exact McNemar p = 0.19; 317 vs 284 discordant documents). The models disagreed on nearly half the documents, so the effect is small relative to this noise. Pooled evidence across all five folds is needed before claiming an improvement.

**I tested the ensemble against the Linear SVC baseline with exact McNemar tests. The improvement is consistent in direction across all four datasets (1.1 to 2.5 points) and statistically supported on the larger test sets, while on the small datasets the tests lack power. I therefore describe the ensemble as a modest improvement and attribute much of it to the selection of strong kernels.**